File: lox/Lox.py
import sys
from Scanner import Scanner

class Lox:
	def __init__(self) -> None:
		self.hadError = False

	def main(self, args: list[str]):
		try:
			if len(args) > 1:
				print("Usage: pyLox [script]")
				exit(64)
			elif len(args) == 1:
				self.runFile(args[0])
			else:
				self.runPrompt()
		except IOError as e:
			print(e)

	def runFile(self, path: str):
		with open(path, 'r') as file:
			self.run(file)

			# Indicate an error has occured
			if self.hadError: exit(65)

	def runPrompt(self):
		command = ""
		while command != "exit;":
			line = input("> ")
			if not line: break
			self.run(line)
			self.hadError = False

	def run(self, source: str):
		scanner = Scanner(source)
		tokens = scanner.scanTokens()

		# Just print the tokens for now
		for t in tokens:
			print(t)

	def error(self, line: int, message: str):
		self.__report(line, "", message)

	def __report(self, line: int, where: str, message: str):
		print("Error:")
		print(f"[line {line}] Error {where}: {message}")
		self.hadError = True


if __name__ == "__main__":
	interpreter = Lox()
	interpreter.main(sys.argv[1:])

File: lox/Lox.cpp
#include <iostream>
#include <string>
#include <fstream>
#include <vector>

#define SCANNER_IMPLEMENTATION
#include "Scanner.h"
// #include "Token.h"

static void run(std::string source) {
    std::cout << "This is yet to come!!!" << source << std::endl;

    // Scanner scanner = new Scanner(source);
    // Token[] tokens = scanner.scanTokens();

    // // Just print the tokens now
    // for (Token token : tokens) {
    //     std::cout << token.toString() << std::endl;
    // }
}

static void error(int line, std::string message) {
    report(line, "", message);
}

static void report(int line, std::string where, std::string message) {
    perror(
        "[line " + (line) //+ "] Error " + where + ": " + message
    );
    hadError = true;
}

static void runFile(std::string path) {
    // Read in the raw file
    std::ifstream bytes (path);
    if (!bytes) throw std::runtime_error("Could not open the file");

    std::string file(
        (std::istreambuf_iterator<char>(bytes)),
        (std::istreambuf_iterator<char>())
    );

    run(file);
}

static void runPrompt() {
    while (true) {
        try {
            std::string line;
            std::cout << "> ";
            std::getline(std::cin, line);
            
            if (line.empty()) return;

            run(line);
        } catch (const std::exception& e) {
            throw;
        }
    }
}

int main(int argc, char* argv[]){
    try {
        if (argc > 2) {
            std::cout << "Usage: slowInterpreter [script]" << std::endl;
            return 64;
        } else if (argc == 2) {
            runFile(argv[1]);
        } else {
            runPrompt();
        }
    } catch (const std::exception& e) {
        std::cout << "Error: " << e.what() << std::endl;
    }

    return 0;
}
